ml_genn/ml_genn/compilers/event_prop_compiler.py

```diff
@@ -0,0 +1,783 @@
+                        example_time=(self.example_timesteps * self.dt),
+                        dynamics="",
+                        transition=""))
+
+                # Prepend code to reset to write spike time to ring buffer
+                model_copy.prepend_reset_code(
+                    neuron_reset.substitute(
+                        max_spikes=self.max_spikes,
+                        write=""))
+            # Otherwise i.e. it's hidden
+            else:
+                # Add additional input variable to receive feedback
+                model_copy.add_additional_input_var("RevISyn", "scalar", 0.0)
+
+                # If neuron model is LIF
+                if isinstance(pop.neuron, LeakyIntegrateFire):
+                    # Check EventProp constraints
+                    if pop.neuron.integrate_during_refrac:
+                        logger.warning("EventProp learning works best with "
+                                       "LIF neurons which do not continue to "
+                                       "integrate in their refractory period")
+                    if pop.neuron.relative_reset:
+                        logger.warning("EventProp learning works best "
+                                       "with LIF neurons with an "
+                                       "absolute reset mechanism")
+
+                    # Get tau_syn from population's incoming connections
+                    tau_syn = _get_tau_syn(pop)
+
+                    # Add parameter with synaptic decay constant
+                    model_copy.add_param("Beta", "scalar",
+                                         np.exp(-self.dt / tau_syn))
+
+                    # Add adjoint state variables
+                    model_copy.add_var("LambdaV", "scalar", 0.0)
+                    model_copy.add_var("LambdaI", "scalar", 0.0)
+
+                    # Add EGP for IMinusV ring variables
+                    model_copy.add_egp("RingIMinusV", "scalar*", 
+                                       np.empty(ring_size, dtype=np.float32))
+
+                    # Add parameter for scaling factor
+                    tau_mem = pop.neuron.tau_mem
+                    model_copy.add_param("A", "scalar", 
+                                         tau_mem / (tau_syn - tau_mem))
+
+                    # On backward pass transition, update LambdaV
+                    transition_code = """
+                        $(LambdaV) += (1.0 / $(RingIMinusV)[ringOffset + $(RingReadOffset)]) * ($(Vthresh) * $(LambdaV) + $(RevISyn));
+                        """
+
+                    # List of variables aside from those in base 
+                    # model we want to reset every batch
+                    additional_reset_vars = [("LambdaV", "scalar", 0.0),
+                                             ("LambdaI", "scalar", 0.0)]
+
+                    # If regularisation is enabled
+                    # **THINK** is this LIF-specific?
```

Comment:
I have never fully worked out the correct math on this but I think it is not LIF-specific.